### PR TITLE
Allow overriding of game factories + blank nextbot interface

### PR DESCRIPTION
**Pull request by Kenzzer**
For a while now, it has been desired by plugin makers to override game's factories. This PR addresses that request, as well as introducing the ability to attach a blank `INextBot` interface to any entity. Furthermore, plugin makers will now be able to use `SetInitialActionFactory` on their factories, in order to override the intention interface of any NextBot entities.

And because a blank `INextBot` interface is now exposed, plugin makers will most certainly need a way to propagate source events, to the NextBot's event capsule. Therefore most of the event natives for `INextBotEventResponder` have been added for plugins to use.

### CODE REVIEW

File: extension/natives/eventresponder.h [hunk 1]
```diff
@@ -25,4 +25,287 @@ EVENTRESPONDERNATIVE(NextContainedResponder)
...
+		return pContext->ThrowNativeError("Entity %d is invalid!", params[2]);
+	}
+	pResponder->OnLandOnGround(entity);
+	return 0;
+}
+
+// TO-DO: OnContact
+
+EVENTRESPONDERNATIVE(OnMoveToSuccess)
+	pResponder->OnMoveToSuccess((Path*)params[2]);
```
Comment (KitRifty) [inline extension/natives/eventresponder.h:51]:
```suggestion
	Path* path = (Path*)params[2];
	if (!path)
	{
		return pContext->ThrowNativeError("Path is NULL");
	}

	pResponder->OnMoveToSuccess(path);
```

File: extension/natives/eventresponder.h [hunk 2]
```diff
@@ -25,4 +25,287 @@ EVENTRESPONDERNATIVE(NextContainedResponder)
...
+	return 0;
+}
+
+EVENTRESPONDERNATIVE(OnMoveToFailure)
+	pResponder->OnMoveToFailure((Path*)params[2], (MoveToFailureType)params[3]);
```
Comment (KitRifty) [inline extension/natives/eventresponder.h:56]:
```suggestion
	Path* path = (Path*)params[2];
	if (!path)
	{
		return pContext->ThrowNativeError("Path is NULL");
	}

	pResponder->OnMoveToFailure(path, (MoveToFailureType)params[3]);
```

File: extension/sourcesdk/NextBot/NextBotInterface.cpp [hunk 3]
```diff
@@ -44,4 +157,186 @@ void INextBot::DebugConColorMsg( NextBotDebugType debugType, const Color &color,
...
+		return true;
+	}
+
+	// TO-DO: fix
+	/*CTeam *team = GetEntity()->GetTeam();
+	if (team && !Q_strnicmp(name, team->GetName(), Q_strlen(name)))
+	{
+		return true;
+	}
+	return false;*/
```
Comment (KitRifty) [inline extension/sourcesdk/NextBot/NextBotInterface.cpp:336]: Need to return something here.

```suggestion
	*/
	return false;
```

**Review by KitRifty — CHANGES_REQUESTED** [review]
Overall solid. I did test and fix this code a little to address several crashes that were happening and should be stable on both platforms. Just some small issues about robustness.

File: extension/natives/eventresponder.h (hunk 1 above)
Comment (Kenzzer) [reply extension/natives/eventresponder.h:51]: Intentional, I was wondering, maybe plugin makers & their custom intention will want to pass null path.

File: extension/natives/eventresponder.h (hunk 2 above)
Comment (Kenzzer) [reply extension/natives/eventresponder.h:56]: Ditto

**Review by KitRifty — APPROVED** [review]

**Pull request merged**